### DISCUSSION TITLE
Explicit move semantics

### DISCUSSION THREAD

**Discussion by kyouko-taiga** [hylo-lang/hylo#1164] [Language design]
We discussed with @dabrahams the idea of requiring moves to be explicit, just like copies. There are actually surprisingly less situations were moves are required than we had initially anticipated. That is because passing arguments to `sink` parameters doesn't require a move at the call site. It only transfers the responsibility to deinitialize the value to the callee, which may transfer it further with a move or any other `sink` operation.

Notable cases where moves are required include when we're returning the value of a local variable from a function. In general, the compiler cannot build that value "in-place" because it may depend on runtime control-flow:

```
fun zero_or_one() -> Int {
  let x = 1
  let y = 0
  return if Bool.random() { x } else { y }
}
```

In this case, the compiler would reject the program and ask for explicit moves:

```
fun zero_or_one() -> Int {
  let x = 1
  let y = 0
  return if Bool.random() { x.move() } else { y.move() }
}
```

Note that there are "better" ways to write this function and avoid moves. Perhaps we could imagine the compiler being able to suggest them in the future:

```
fun zero_or_one() -> Int {
  if Bool.random() { 1 } else { 0 }
}
```

Making move explicit removes some of the perceived "magic" that may surround conversions. For example, given `x: T` the expression `x as Optional<_>` must consume `x`, causing a move. If `T` is not movable, this expression will fail but `T() as Optional<_>` won't because here the compiler can initialize the instance in place.

With explicit moves, the user will still be confronted to the difference between in-place initialization and move-initialization, but at least we'll have a way to explain how things work at the source level. Currently, the move required in `x as Optional<_>` is inserted automatically by the compiler.

The implementation of `move` is probably not representable in source because we don't have a way to name the return storage of a function. However, we can supply it as an extension of `Movable` with a compiler-generated definition:

```
public extension Movable {
  public fun move() sink -> { /* compiler-magic */ }
}
```

**Comment by tothambrus11** [comment]
It's an interesting observation. I'm not sure that explicitness has much benefit, as moves are usually cheap. For the record: in hylo-new, we now use implicit moves and explicit copies by default.

**Reply by dabrahams** [reply]
As in the old hylo, by design
